sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -80,28 +105,57 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
    * transformExpressionsDown or transformExpressionsUp should be used.
    *
    * @param rule the rule to be applied to every expression in this operator.
+   * @param cond  a Lambda expression to prune tree traversals. If `cond.apply` returns false
+   *              on an expression T, skips processing T and its subtree; otherwise, processes
+   *              T and its subtree recursively.
+   * @param ruleId is a unique Id for `rule` to prune unnecessary tree traversals. When it is
+   *        RuleId.UnknownId, no pruning happens. Otherwise, if `rule`(with id `ruleId`) has been
+   *        marked as in effective on an expression T, skips processing T and its subtree. Do not
+   *        pass it if the rule is not purely functional and reads a varying initial state for
+   *        different invocations.
    */
-  def transformExpressions(rule: PartialFunction[Expression, Expression]): this.type = {
-    transformExpressionsDown(rule)
+  def transformExpressions(rule: PartialFunction[Expression, Expression],
+    cond: TreePatternBits => Boolean = AlwaysProcess.fn,
+    ruleId: Int = RuleIdCollection.UnknownId): this.type = {
```

Comment:
Why not just pass in the Rule itself or some RuleId trait? That way you make it more explicit what you expect here.

Reply:
Yes, that's definitely a better approach -- if I implement it from scratch, I'd do that way.
However, I don't have a good idea to get existing code like "plan transform {...}" compile if change "rule: PartialFunction" to "rule: Rule".  If we change every call sites of transform, the PR might go too large.

Reply:
I was more referring that instead of an ID we should pass in the Rule or some identifier instead of an int.

Passing in the rule would mean that we'd have the transformation logic in a different place. While that is definitely interesting, it is indeed a massive change.

Reply:
>> some identifier instead of an int

Makes sense. I'll add a RuleId type to wrap the Integer.

Reply:
Done. Now pass an RuleId object (with a range requirement) instead of a raw int.